src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs

```diff
@@ -0,0 +1,536 @@
+{
+    public interface IChainedHeaderValidator
+    {
+        void Validate(ChainedHeader chainedHeader);
+    }
+
+    /// <summary>
+    /// A tree structure of <see cref="ChainedHeader "/> elements.
+    /// </summary>
+    /// <remarks>
+    /// A blockchain can have multiple versions of the chain, with only one being the longest chain (or the chain with most proof).
+    /// While this is not ideal it can happen naturally or maliciously, to be able to find the best chain we have to keep track of all chains we discover.
+    /// Only one chain will represent the tip of the blockchain. 
+    /// </remarks>
+    public sealed class ChainedHeaderTree
+    {
+        private readonly Network network;
+        private readonly IChainedHeaderValidator chainedHeaderValidator;
+        private readonly ILogger logger;
+        private readonly ICheckpoints checkpoints;
+        private readonly IChainState chainState;
+        private readonly ConsensusSettings consensusSettings;
+
+        /// <summary>A special peer identifier that represents our local node.</summary>
+        internal const int LocalPeerId = -1;
+
+        /// <summary>A list of headers that represent the tips of peers.</summary>
+        private readonly Dictionary<uint256, HashSet<int>> peerIdsByTipHash;
+
+        /// <summary>A list of peer identifiers that map to the peers tip has.</summary>
+        private readonly Dictionary<int, uint256> peerTipsByPeerId;
+
+        /// <summary>An indexed collection of <see cref="ChainedHeader"/> that represents a tree of chains.</summary>
+        private readonly Dictionary<uint256, ChainedHeader> chainedHeadersByHash;
+
+        internal Dictionary<uint256, ChainedHeader> GetChainedHeadersByHash => this.chainedHeadersByHash;
+        internal Dictionary<uint256, HashSet<int>> GetPeerIdsByTipHash => this.peerIdsByTipHash;
+        internal Dictionary<int, uint256> GetPeerTipsByPeerId => this.peerTipsByPeerId;
+
+        public ChainedHeaderTree(
+            Network network, 
+            ILoggerFactory loggerFactory, 
+            IChainedHeaderValidator chainedHeaderValidator, 
+            ICheckpoints checkpoints, 
+            IChainState chainState, 
+            ConsensusSettings consensusSettings)
+        {
+            this.network = network;
+            this.chainedHeaderValidator = chainedHeaderValidator;
+            this.checkpoints = checkpoints;
+            this.chainState = chainState;
+            this.consensusSettings = consensusSettings;
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+
+            this.peerTipsByPeerId = new Dictionary<int, uint256>();
+            this.peerIdsByTipHash = new Dictionary<uint256, HashSet<int>>();
+            this.chainedHeadersByHash = new Dictionary<uint256, ChainedHeader>();
+        }
+
+        /// <summary>
+        /// Initialize the tree with a <see cref="ChainedHeader"/> that is the tip of our best known consensus chain.
```

Comment:
Maybe `Initialize the tree with consensus tip`?

Reply:
Agree